feat: introduce generated_file_test

alexeagle opened the pull request:
This is a new public API for checking in golden or snapshot files.
If you used the private golden_file_test previously, you need to rename actual to generated, golden to src, and golden_debug to src_dbg

Fixes #1893

jbedard commented on internal/generated_file_test/bin.js:
```diff
@@ -0,0 +1,48 @@
+    return 0;
+  }
+  if (mode === '--verify') {
+    const unidiff = require('unidiff');
```
Nothing needs to be done to ensure `unidiff` is available?

I know the `@bazel/*` npm packages use `peerDependencies`, but how does it work for rules built directly into rules_nodejs?

jbedard replied:
I see `@npm//unidiff` is passed into the `nodejs_test|binary(data)`, but is that the consumers `@npm//` and we assume the package exists?

alexeagle replied:
Immediate answer is this needs some e2e test to show how it's used outside our workspace.
Yes I think that will show it's broken like you say. This rule has to be in an npm package or else unidiff has to be vendored.

jbedard replied:
That could always be a followup, wouldn't change what you've done.

alexeagle commented:
@jbedard I added a second commit where I run rollup on the bin.js so unidiff gets inlined into the code. Added a test by simplifying the parcel example with a golden file rather than a spec, to prove that it works.
